### PR TITLE
[Merged by Bors] - Data.Nat.Choose.Basic: Register a faster implementation

### PR DESCRIPTION
**Pull request by nomeata**
---

Discussed in https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/memoziation.20of.20strong.20recursion.20on.20nat/near/357298523

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Nat/Choose/Basic.lean [hunk 1]
```diff
@@ -273,6 +273,12 @@ theorem choose_eq_descFactorial_div_factorial (n k : ℕ) : n.choose k = n.descF
   exact (Nat.mul_div_cancel' <| factorial_dvd_descFactorial _ _).symm
 #align nat.choose_eq_desc_factorial_div_factorial Nat.choose_eq_descFactorial_div_factorial
 
+/-- A faster implementation of `choose`, to be used during bytecode evaluation -/
```
[the remark this answers is not in the document]
Comment (nomeata) [reply Mathlib/Data/Nat/Choose/Basic.lean:276]:
```suggestion
/-- A faster implementation of `choose`, to be used during bytecode evaluation and in compiled code. -/
```

File: Mathlib/Data/Nat/Choose/Basic.lean [hunk 2]
```diff
@@ -273,6 +273,12 @@ theorem choose_eq_descFactorial_div_factorial (n k : ℕ) : n.choose k = n.descF
   exact (Nat.mul_div_cancel' <| factorial_dvd_descFactorial _ _).symm
 #align nat.choose_eq_desc_factorial_div_factorial Nat.choose_eq_descFactorial_div_factorial
 
+/-- A faster implementation of `choose`, to be used during bytecode evaluation and in compiled code. -/
```
Comment (nomeata) [inline Mathlib/Data/Nat/Choose/Basic.lean:276]:
```suggestion
/-- A faster implementation of `choose`, to be used during bytecode evaluation
and in compiled code. -/
```

**Review by mattrobball — APPROVED** [review]
This is way, way faster. Nice!

@nomeata You may want to add the `awaiting-review` tag and probably the `new-feature` tag if you want it be reviewed for merging.

File: Mathlib/Data/Nat/Choose/Basic.lean [hunk 3]
```diff
@@ -273,6 +273,13 @@ theorem choose_eq_descFactorial_div_factorial (n k : ℕ) : n.choose k = n.descF
   exact (Nat.mul_div_cancel' <| factorial_dvd_descFactorial _ _).symm
 #align nat.choose_eq_desc_factorial_div_factorial Nat.choose_eq_descFactorial_div_factorial
 
+/-- A faster implementation of `choose`, to be used during bytecode evaluation
+and in compiled code. -/
+def fast_choose n k := Nat.descFactorial n k / Nat.factorial k
+@[csimp] lemma choose_eq_fast_choose : Nat.choose = fast_choose :=
```
Comment (kmill) [inline Mathlib/Data/Nat/Choose/Basic.lean:279]:
```suggestion

@[csimp] lemma choose_eq_fast_choose : Nat.choose = fast_choose :=
```

**Comment by kmill** [comment]
Thanks!

bors r+

**Pull request closed without merging**